airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/SSHTunnel.java

```diff
@@ -0,0 +1,277 @@
+  private final String method;
+  private final String host;
+  private final String tunnelSshPort;
+  private final String user;
+  private final String sshkey;
+  private final String password;
+  private final String remoteDatabaseHost;
+  private final String remoteDatabasePort;
+  private final String tunnelDatabasePort;
+
+  private SSHTunnel tunnelConfig = null;
+  private SshClient sshclient = null;
+  private ClientSession tunnelSession = null;
+
+  public SSHTunnel(String method,
+                   String host,
+                   String tunnelSshPort,
+                   String user,
+                   String sshkey,
+                   String password,
+                   String remoteDatabaseHost,
+                   String remoteDatabasePort,
+                   String tunnelDatabasePort) {
+    if (method == null) {
+      this.method = "NO_TUNNEL";
+    } else {
+      this.method = method;
+    }
+    this.host = host;
+    this.tunnelSshPort = tunnelSshPort;
+    this.user = user;
+    this.sshkey = sshkey;
+    this.password = password;
+    this.remoteDatabaseHost = remoteDatabaseHost;
+    this.remoteDatabasePort = remoteDatabasePort;
+    this.tunnelDatabasePort = tunnelDatabasePort;
+  }
+
+  public static SSHTunnel getInstance(JsonNode config) {
+    JsonNode ourConfig = config.get("tunnel_method");
+    SSHTunnel sshconfig = new SSHTunnel(
+        getConfigValueOrNull(ourConfig, "tunnel_method"),
+        getConfigValueOrNull(ourConfig, "tunnel_host"),
+        getConfigValueOrNull(ourConfig, "tunnel_ssh_port"),
+        getConfigValueOrNull(ourConfig, "tunnel_username"),
+        getConfigValueOrNull(ourConfig, "tunnel_usersshkey"),
+        getConfigValueOrNull(ourConfig, "tunnel_userpass"),
+        getConfigValueOrNull(ourConfig, "tunnel_db_remote_host"),
+        getConfigValueOrNull(ourConfig, "tunnel_db_remote_port"),
+        getConfigValueOrNull(ourConfig, "tunnel_localport"));
+    return sshconfig;
+  }
+
+  static String getConfigValueOrNull(JsonNode config, String key) {
+    return config != null && config.has(key) ? config.get(key).asText() : null;
+  }
+
+  /**
+   * Starts an ssh session; wrap this in a try-finally and use closeTunnel() to close it.
+   *
+   * @throws IOException
+   */
+  public void openTunnelIfRequested() throws IOException {
+    if (shouldTunnel()) {
+      try {
+        throw new Exception("Troubleshooting JENNY");
+      } catch (Exception e) {
+        LOGGER.error("Troubleshooting! ", e);
+      }
+      if (tunnelSession != null || sshclient != null) {
+        throw new RuntimeException("SSH Tunnel was requested to be opened while it was already open.  This is a coding error.");
+      }
+      sshclient = createClient();
+      tunnelSession = openTunnel(sshclient);
+    }
+  }
+
+  /**
+   * Closes a tunnel if one was open, and otherwise doesn't do anything (safe to run).
+   */
+  public void closeTunnel() {
```

Comment:
This could actually be `close()` and the class could be `AutoClosable` so we could do this with a try-with-resources block instead of a try-finally block.